webapi/middleware.go

```diff
@@ -162,28 +163,68 @@ func broadcastTicket() gin.HandlerFunc {
 		}
 
 		dcrdClient := c.MustGet("DcrdClient").(*rpc.DcrdRPC)
+		var e *wsrpc.Error
+
+		if request.ParentHex != nil {
+			parentTx, err := decodeTransaction(*request.ParentHex)
+			if err != nil {
+				log.Errorf("%s: Failed to decode parent hex (ticketHash=%s): %v", funcName, request.TicketHash, err)
+				sendErrorWithMsg("cannot decode parent hex", errBadRequest, c)
+				return
+			}
+			parentHash := parentTx.TxHash()
+			var found bool
+			for _, txIn := range msgTx.TxIn {
+				if !txIn.PreviousOutPoint.Hash.IsEqual(&parentHash) {
+					continue
+				}
+				found = true
+				break
+			}
+			if !found {
+				log.Errorf("%s: Invalid ticket parent (ticketHash=%s)", funcName, request.TicketHash)
+				sendErrorWithMsg("invalid ticket parent", errBadRequest, c)
+				return
+			}
+			log.Debugf("%s: Broadcasting parent tx %s (ticketHash=%s)", funcName, parentHash, request.TicketHash)
+			err = dcrdClient.SendRawTransaction(request.TicketHex)
+			switch {
+			case err == nil:
+				// Sent
+			case errors.As(err, &e) && e.Code == rpc.ErrRPCDuplicateTx:
+				// dcrd already has the parent
+			default:
+				log.Errorf("%s: dcrd.SendRawTransaction for parent tx failed (ticketHash=%s): %v",
+					funcName, request.TicketHash, err)
+				sendError(errCannotBroadcastTicket, c)
+				return
+			}
+		}
 
 		// Use GetRawTransaction to check if local dcrd already knows this
 		// ticket.
 		_, err = dcrdClient.GetRawTransaction(request.TicketHash)
-		if err == nil {
+		switch {
+		case err == nil:
 			// No error means dcrd already knows the ticket, we are done here.
 			return
-		}
-
-		// ErrNoTxInfo means local dcrd is not aware of the ticket. We have the
-		// hex, so we can broadcast it here.
-		var e *wsrpc.Error
-		if errors.As(err, &e) && e.Code == rpc.ErrNoTxInfo {
+		case errors.As(err, &e) && e.Code == rpc.ErrNoTxInfo:
+			// ErrNoTxInfo means local dcrd is not aware of the ticket. We have the
+			// hex, so we can broadcast it here.
 			log.Debugf("%s: Broadcasting ticket (ticketHash=%s)", funcName, request.TicketHash)
 			err = dcrdClient.SendRawTransaction(request.TicketHex)
-			if err != nil {
+			switch {
```

Comment:
Duplicate tx errors already handled in the RPC client https://github.com/decred/vspd/blob/0f657eb4fe6b143d757a5cd2936b7d04cef6849e/rpc/dcrd.go#L121-L158